From f1ce1d0e4a002aa621fd894c9c4eaa75939a9ea2 Mon Sep 17 00:00:00 2001
From: Ayush Dey <deyayush6@gmail.com>
Date: Mon, 11 Aug 2025 00:23:33 +0530
Subject: [PATCH 13/27] vlc_stream: add STREAM_CAN_DOWNLOAD control query

To identify access modules that will support file downloading
via the future LibVLC downloader API. This query is intended
for a limited set of modules where download capability is applicable.
---
 include/vlc_stream.h         | 8 ++++++++
 modules/access/dsm/access.c  | 1 +
 modules/access/file.c        | 5 +++++
 modules/access/ftp.c         | 1 +
 modules/access/http.c        | 1 +
 modules/access/http/access.c | 1 +
 modules/access/nfs.c         | 1 +
 modules/access/samba.c       | 1 +
 modules/access/sftp.c        | 1 +
 modules/access/smb2.c        | 1 +
 10 files changed, 21 insertions(+)

diff --git a/include/vlc_stream.h b/include/vlc_stream.h
index bb5bdc61e3..b0606a799f 100644
--- a/include/vlc_stream.h
+++ b/include/vlc_stream.h
@@ -255,6 +255,7 @@ enum stream_query_e
     STREAM_CAN_FASTSEEK,                    /**< arg1=(bool *) res=cannot fail */
     STREAM_CAN_PAUSE,                       /**< arg1=(bool *) res=cannot fail */
     STREAM_CAN_CONTROL_PACE,                /**< arg1=(bool *) res=cannot fail */
+    STREAM_CAN_DOWNLOAD,                    /**< arg1=(bool *) res=can fail */
     /* */
     STREAM_GET_SIZE=6,                      /**< arg1=(uint64_t *) res=can fail */
     STREAM_GET_MTIME,                       /**< arg1=(uint64_t *) res=can fail
@@ -468,6 +469,13 @@ VLC_USED static inline bool vlc_stream_CanPace(stream_t *s)
     return can_control_pace;
 }
 
+VLC_USED static inline bool vlc_stream_CanDownload(stream_t *s)
+{
+    bool can_download = false;
+    vlc_stream_Control(s, STREAM_CAN_DOWNLOAD, &can_download);
+    return can_download;
+}
+
 VLC_USED static inline vlc_tick_t vlc_stream_GetPtsDelay(stream_t *s)
 {
     vlc_tick_t pts_delay;
diff --git a/modules/access/dsm/access.c b/modules/access/dsm/access.c
index 8e2c36a216..b343f1580f 100644
--- a/modules/access/dsm/access.c
+++ b/modules/access/dsm/access.c
@@ -733,6 +733,7 @@ static int Control( stream_t *p_access, int i_query, va_list args )
     case STREAM_CAN_SEEK:
     case STREAM_CAN_PAUSE:
     case STREAM_CAN_CONTROL_PACE:
+    case STREAM_CAN_DOWNLOAD:
         *va_arg( args, bool* ) = true;
         break;
 
diff --git a/modules/access/file.c b/modules/access/file.c
index 35aac2db39..8d57fd3d74 100644
--- a/modules/access/file.c
+++ b/modules/access/file.c
@@ -361,6 +361,11 @@ static int FileControl( stream_t *p_access, int i_query, va_list args )
             *pb_bool = p_sys->b_pace_control;
             break;
 
+        case STREAM_CAN_DOWNLOAD:
+            pb_bool = va_arg( args, bool * );
+            *pb_bool = true;
+            break;
+
         case STREAM_GET_SIZE:
         case STREAM_GET_MTIME:
         {
diff --git a/modules/access/ftp.c b/modules/access/ftp.c
index c1ce2ffef7..a69c33d47f 100644
--- a/modules/access/ftp.c
+++ b/modules/access/ftp.c
@@ -1342,6 +1342,7 @@ static int Control( stream_t *p_access, int i_query, va_list args )
             *pb_bool = true;    /* FIXME */
             break;
         case STREAM_CAN_CONTROL_PACE:
+        case STREAM_CAN_DOWNLOAD:
             pb_bool = va_arg( args, bool * );
             *pb_bool = true;    /* FIXME */
             break;
diff --git a/modules/access/http.c b/modules/access/http.c
index 4278f51fba..5159d58f1e 100644
--- a/modules/access/http.c
+++ b/modules/access/http.c
@@ -570,6 +570,7 @@ static int Control( stream_t *p_access, int i_query, va_list args )
             break;
         case STREAM_CAN_PAUSE:
         case STREAM_CAN_CONTROL_PACE:
+        case STREAM_CAN_DOWNLOAD:
             pb_bool = va_arg( args, bool* );
             *pb_bool = true;
             break;
diff --git a/modules/access/http/access.c b/modules/access/http/access.c
index 4302c0d455..ffc38b749e 100644
--- a/modules/access/http/access.c
+++ b/modules/access/http/access.c
@@ -82,6 +82,7 @@ static int FileControl(stream_t *access, int query, va_list args)
 
         case STREAM_CAN_PAUSE:
         case STREAM_CAN_CONTROL_PACE:
+        case STREAM_CAN_DOWNLOAD:
             *va_arg(args, bool *) = true;
             break;
 
diff --git a/modules/access/nfs.c b/modules/access/nfs.c
index 3510c3fe1f..5083bb2a32 100644
--- a/modules/access/nfs.c
+++ b/modules/access/nfs.c
@@ -289,6 +289,7 @@ FileControl(stream_t *p_access, int i_query, va_list args)
 
         case STREAM_CAN_PAUSE:
         case STREAM_CAN_CONTROL_PACE:
+        case STREAM_CAN_DOWNLOAD:
             *va_arg(args, bool *) = true;
             break;
 
diff --git a/modules/access/samba.c b/modules/access/samba.c
index 9ee4bea188..5e9330dfdd 100644
--- a/modules/access/samba.c
+++ b/modules/access/samba.c
@@ -217,6 +217,7 @@ static int Control( stream_t *p_access, int i_query, va_list args )
     case STREAM_CAN_SEEK:
     case STREAM_CAN_PAUSE:
     case STREAM_CAN_CONTROL_PACE:
+    case STREAM_CAN_DOWNLOAD:
         *va_arg( args, bool* ) = true;
         break;
 
diff --git a/modules/access/sftp.c b/modules/access/sftp.c
index 089bf3d7e3..dcb90fb85d 100644
--- a/modules/access/sftp.c
+++ b/modules/access/sftp.c
@@ -641,6 +641,7 @@ static int Control( stream_t* p_access, int i_query, va_list args )
 
     case STREAM_CAN_PAUSE:
     case STREAM_CAN_CONTROL_PACE:
+    case STREAM_CAN_DOWNLOAD:
         pb_bool = va_arg( args, bool * );
         *pb_bool = true;
         break;
diff --git a/modules/access/smb2.c b/modules/access/smb2.c
index 793fcf92cf..d8c61c691a 100644
--- a/modules/access/smb2.c
+++ b/modules/access/smb2.c
@@ -364,6 +364,7 @@ FileControl(stream_t *access, int i_query, va_list args)
 
         case STREAM_CAN_PAUSE:
         case STREAM_CAN_CONTROL_PACE:
+        case STREAM_CAN_DOWNLOAD:
             *va_arg(args, bool *) = true;
             break;
 
-- 
2.50.1 (Apple Git-155)

